### PR TITLE
Work on executors

### PR DESCRIPTION
**Pull request by artem-v**

### CODE REVIEW

File: cluster/src/main/java/io/scalecube/cluster/gossip/GossipProtocol.java [hunk 1]
```diff
@@ -86,17 +93,8 @@ public GossipProtocol(String memberId, ITransport transport, GossipProtocolConfi
...
-  private ThreadFactory createThreadFactory(String namingFormat) {
-    return new ThreadFactoryBuilder().setNameFormat(namingFormat)
-        .setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
-          @Override
-          public void uncaughtException(Thread thread, Throwable ex) {
-            LOGGER.error("Unhandled exception: {}", ex, ex);
-          }
-        }).setDaemon(true).build();
+    this.executor = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setNameFormat("sc-gossip-%s").setDaemon(true).build());
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/gossip/GossipProtocol.java]: What about uncaught exception handler? Isn't it needed?

File: cluster/src/main/java/io/scalecube/cluster/fdetector/FailureDetector.java [hunk 2]
```diff
@@ -343,4 +334,63 @@ public Boolean call(Message message) {
...
+      Address target = data.getOriginalIssuer();
+      String correlationId = message.correlationId();
+      PingData originalAckData = new PingData(target, data.getTo());
+      Message originalAckMessage =
+          Message.withData(originalAckData).qualifier(ACK).correlationId(correlationId).build();
+      transport.send(target, originalAckMessage);
+    }
+  }
+
+  private class FailureDetectorProtocolRunnable implements Runnable {
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/fdetector/FailureDetector.java]: Maybe just PingTask or so...

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java [hunk 3]
```diff
@@ -81,64 +87,29 @@ public Boolean call(Message message) {
...
+  private Subscriber<ClusterMembershipData> onGossipRequestSubscriber;
+  private Subject<ClusterMember, ClusterMember> subject =
+      new SerializedSubject<>(PublishSubject.<ClusterMember>create());
+
+  // Scheduled
+
+  private final Scheduler scheduler;
+  private ScheduledFuture<?> executorTask;
+  private final ScheduledExecutorService executor;
+  private final ScheduledExecutorService suspectedMembersTimer;
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: Why another executor is needed for it? Can't we use just executor?

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java [hunk 4]
```diff
@@ -68,11 +71,14 @@
   private static final Func1<Message, Boolean> GOSSIP_MEMBERSHIP_FILTER = new Func1<Message, Boolean>() {
     @Override
     public Boolean call(Message message) {
-      Object data = message.data();
-      return data != null && ClusterMembershipData.class.equals(data.getClass());
+      return message.data() != null && ClusterMembershipData.class.isAssignableFrom(message.data().getClass());
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: Why "isAssignableFrom"? We don't need to support superclasses or super interfaces here.

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java [hunk 5]
```diff
@@ -463,11 +446,91 @@ public boolean isLocalMember(ClusterMember member) {
...
+    public void call(ClusterMembershipData data) {
+      List<ClusterMember> updates = membership.merge(data);
+      if (!updates.isEmpty()) {
+        LOGGER.debug("Received gossip, updates: {}", updates);
+        processUpdates(updates, false/* spread gossip */);
+      }
+    }
+  }
+
+  private class MembershipProtocolRunnable implements Runnable {
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: Maybe just SyncTask

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java [hunk 6]
```diff
@@ -418,29 +399,31 @@ private void processUpdates(List<ClusterMember> updates, boolean spreadGossip) {
     for (final ClusterMember member : updates) {
       LOGGER.debug("Member {} became {}", member.address(), member.status());
       switch (member.status()) {
-        case SUSPECTED:
+        case SUSPECTED: {
```
Comment (antonkharenko) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: why we need brackets here?

File: cluster/src/main/java/io/scalecube/cluster/gossip/GossipProtocol.java (hunk 1 above)
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/gossip/GossipProtocol.java]: @antonkharenko correct, not needed.

File: cluster/src/main/java/io/scalecube/cluster/fdetector/FailureDetector.java (hunk 2 above)
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/fdetector/FailureDetector.java]: @antonkharenko ok

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java (hunk 4 above)
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: @antonkharenko ok

File: cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java (hunk 6 above)
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/ClusterMembership.java]: @antonkharenko removed curly braces. take a look at code changes in this block again, I changed ``case TRUSTED:``  block.

**Pull request merged**